### PR TITLE
NIFI-8248 Modify PutAzureDataLakeStorage processor to use temp file i…

### PR DESCRIPTION
**Pull request by timeabarna**
# Summary

[NIFI-8248](https://issues.apache.org/jira/browse/NIFI-8248) enables the usage of a temporary file instead of immediate overwrite.

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 8
  - [ ] JDK 11
  - [ ] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

### CODE REVIEW

**Comment by timeabarna** [comment]
Hello @jfrazee and @exceptionfactory,

I've managed to get to modifying PitAzureDataLake processor upload.
Last year we had a long discussion how different service like Spark or HDFS could be impacted by this change.
Based on that discussion can be found [here](https://github.com/apache/nifi/pull/5098) the following design decisions have been made:

List processor: 
     - A boolean property has been created to turn off showing temp directory listing. If the list processor would hide files starting with '.' or '_' set to default true, users listing hidden files now would be impacted. If the default value would be false, temporary files would appear in listing which may again be an impact for users.

Put processor:
    - A property for temp directory location has been created.
    - Temp files created in a static folder under a generated UUID with the existing file name.
    
When you have some time can you please check? 
Thanks in advance for your help.

**Comment by timeabarna** [comment]
Hello @jfrazee and @exceptionfactory,

When you have some time can you please check this PR?
Thanks in advance for your help.

**Comment by timeabarna** [comment]
Hello @jfrazee and @exceptionfactory,

When you have some time can you please check this PR?
Thanks in advance for your help.

**Comment by exceptionfactory** [comment]
Thanks for the reminder @timeabarna, will plan on reviewing this soon.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for putting together this new PR @timeabarna! The general approach makes sense, and it is helpful to consider the temporary file handling in the context of both the List and Put Processors.

I noted some questions and suggestions on the property naming in particular. It was a little hard to follow some of the naming, and I think it is important to keep that clear for end users. The general approach makes sense.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java [hunk 1]
```diff
@@ -83,11 +89,23 @@ public class PutAzureDataLakeStorage extends AbstractAzureDataLakeStorageProcess
             .allowableValues(FAIL_RESOLUTION, REPLACE_RESOLUTION, IGNORE_RESOLUTION)
             .build();
 
+    public static final PropertyDescriptor TEMP_FILE_DIRECTORY_PATH = new PropertyDescriptor.Builder()
+            .name("temp-file-directory-path")
+            .displayName("Temp File Directory Path")
```
[the remark this answers is not in the document]
Comment (timeabarna) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java:94]: You are correct, this is a prefix. I've modified the wording, it can be tweaked further if necessary.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/AbstractAzureDataLakeStorageProcessor.java [hunk 2]
```diff
@@ -101,6 +101,8 @@ public abstract class AbstractAzureDataLakeStorageProcessor extends AbstractProc
             REL_FAILURE
     )));
 
+    public static final String TEMP_FILE_DIRECTORY = "_$azuretempdirectory$";
```
[the remark this answers is not in the document]
Comment (timeabarna) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/AbstractAzureDataLakeStorageProcessor.java:104]: I used the name can be found in this review [comment] https://github.com/apache/nifi/pull/5098#discussion_r675242859). My assumption was that $ character may be means something for either Hadoop or Azure, it seems I was wrong. I can rename the directory to '_nifitempdirectory'. What do you think?

**Comment by timeabarna** [comment]
Hello @exceptionfactory,

Thank you very much for your review. I've modified the code based on your recommendations.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/ListAzureDataLakeStorage.java [hunk 3]
```diff
@@ -129,13 +130,23 @@ public class ListAzureDataLakeStorage extends AbstractListAzureProcessor<ADLSFil
             .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
             .build();
 
+    public static final PropertyDescriptor INCLUDE_TEMPORARY_FILES = new PropertyDescriptor.Builder()
+            .name("include-temporary-files")
+            .displayName("include temporary files")
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/ListAzureDataLakeStorage.java:135]: Display Names should be capitalized:
```suggestion
            .displayName("Include Temporary Files")
```

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java [hunk 4]
```diff
@@ -83,11 +88,23 @@ public class PutAzureDataLakeStorage extends AbstractAzureDataLakeStorageProcess
...
 
+    public static final PropertyDescriptor BASE_TEMPORARY_PATH = new PropertyDescriptor.Builder()
+            .name("base-temporary-path")
+            .displayName("Base Temporary Path")
+            .description("The Path where the temporary directory will be created. The Path name cannot contain a leading '/'." +
+                    " The root directory can be designated by the empty string value. Non-existing directories will be created." +
+                    "The Temporary File Directory name is " + TEMP_FILE_DIRECTORY)
+            .defaultValue("")
+            .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+            .addValidator(new DirectoryValidator("Temp File Directory"))
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java:99]: Should this label be adjusted to match the property name?
```suggestion
            .addValidator(new DirectoryValidator("Base Temporary Path"))
```

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/AbstractAzureDataLakeStorageProcessor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/AbstractAzureDataLakeStorageProcessor.java:104]: Thanks for the reply and reference to the previous discussion. Using `_nifitempdirectory` sounds good.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for making the adjustments @timeabarna, the rewording is helpful.

I noted a couple more very minor wording adjustments, but otherwise this seems close to completion.

Do you have any additional feedback @jfrazee?

Pending these final updates, this should be ready to merge soon.

**Comment by timeabarna** [comment]
Thanks @exceptionfactory, I've updated the PR.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/resources/docs/org.apache.nifi.processors.azure.storage.PutAzureDataLakeStorage/additionalDetails.html [hunk 5]
```diff
@@ -28,27 +28,29 @@
...
-     <li>Processors with "ignore" conflict resolution strategy will be directed to "Success" relationship.</li>
+    <li>Processors with "ignore" conflict resolution strategy will be directed to "Success" relationship.</li>
     <li>Processors with "replace" conflict resolution strategy:</li>
 
     <ol>
-        <li>An empty file overwrites the existing file, the original file is lost.</li>
-        <li>Content is appended to file.</li>
-        <li>In case append failure the file is deleted.</li>
-        <li>In case file deletion failure the empty file remains on the server.</li>
+        <li>A temporary file is created with random prefix under the given path in '_$azuretempdirectory$'.</li>
```
Comment (jfrazee) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/resources/docs/org.apache.nifi.processors.azure.storage.PutAzureDataLakeStorage/additionalDetails.html:49]:
```suggestion
        <li>A temporary file is created with random prefix under the given path in '_$nifitempdirectory$'.</li>
```

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/resources/docs/org.apache.nifi.processors.azure.storage.PutAzureDataLakeStorage/additionalDetails.html [hunk 6]
```diff
@@ -28,27 +28,29 @@
...
 
-<h4>New file</h4>
+<h4>New file upload</h4>
 
 <ol>
-    <li>An empty file is created.</li>
-    <li>Content is appended to file.</li>
-    <li>In case append failure the file is deleted.</li>
-    <li>In case file deletion failure the empty file remains on the server.</li>
+    <li>A temporary file is created with random prefix under the given path in '_$azuretempdirectory$'.</li>
```
Comment (jfrazee) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/resources/docs/org.apache.nifi.processors.azure.storage.PutAzureDataLakeStorage/additionalDetails.html:34]: I think this changed to `_nifitempdirectory` in the final commits.
```suggestion
    <li>A temporary file is created with random prefix under the given path in '_$nifitempdirectory$'.</li>
```

**Review by jfrazee — CHANGES_REQUESTED** [review]
I did the full build, ran the ITs, tested the upgrade path, ran concurrent Puts with 100MB files, and List with and without temp files enabled. Everything worked as expected.

I think the temp dir name was changed during the reviews and is wrong in the additionalDetails.html now. If we get that fixed I'm +1 on this.

Thanks for all the work and patience! I think this is a really important improvement.

**Comment by timeabarna** [comment]
Good catch @jfrazee, I totally overlooked this. Thank you very much for your help.

**Review by exceptionfactory — APPROVED** [review]
Thanks again for working on this issue and addressing the feedback @timeabarna! The latest version looks good.

@jfrazee Feel free to merge when you approve of the latest adjustments.
